### PR TITLE
Handle response content length mismatches (#175)

### PR DESCRIPTION
**Pull request by cesarbs**
#175 

- [x] TODO: `InvalidOperationException` might not be the right exception type for this
- [x] TODO: is logging this as application errors OK or should we add new event IDs for response length mismatches?

@halter73 @Tratcher @mikeharder @davidfowl

### CODE REVIEW

File: src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/Frame.cs [hunk 1]
```diff
@@ -598,6 +604,34 @@ public Task WriteAsync(ArraySegment<byte> data, CancellationToken cancellationTo
...
+        }
+
+        protected void TryInitializeResponseContentLength()
+        {
+            if (!_responseContentLength.HasValue)
+            {
+                var responseHeaders = FrameResponseHeaders;
+                if (responseHeaders != null && responseHeaders.HasContentLength)
+                {
+                    _responseContentLength = int.Parse(responseHeaders.HeaderContentLength.ToString());
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/Frame.cs]: TryParse, culture, format

File: src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/Frame.cs [hunk 2]
```diff
@@ -598,6 +604,34 @@ public Task WriteAsync(ArraySegment<byte> data, CancellationToken cancellationTo
...
+
+            if (_responseContentLength.HasValue && _responseBytesWritten > _responseContentLength.Value)
+            {
+                ThrowContentLengthMismatchException();
+            }
+        }
+
+        protected void TryInitializeResponseContentLength()
+        {
+            if (!_responseContentLength.HasValue)
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/Frame.cs]: If there's no content-length header this will re-run for every write. You need a flag to only run it once, or is HasContentLength really that fast?

File: src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/FrameOfT.cs [hunk 3]
```diff
@@ -92,6 +92,13 @@ public class Frame<TContext> : Frame
                         try
                         {
                             await _application.ProcessRequestAsync(context).ConfigureAwait(false);
+
+                            TryInitializeResponseContentLength();
+
+                            if (_responseContentLength.HasValue && _responseBytesWritten < _responseContentLength.Value)
+                            {
+                                ThrowContentLengthMismatchException();
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/FrameOfT.cs]: This should be a different message "Too few bytes written".

File: src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/Frame.cs [hunk 4]
```diff
@@ -598,6 +604,34 @@ public Task WriteAsync(ArraySegment<byte> data, CancellationToken cancellationTo
...
+                {
+                    _responseContentLength = int.Parse(responseHeaders.HeaderContentLength.ToString());
+                }
+            }
+        }
+
+        protected void ThrowContentLengthMismatchException()
+        {
+            _keepAlive = false;
+            throw new InvalidOperationException($"Response content length mismatch: wrote {_responseBytesWritten} bytes to {_responseContentLength.Value}-byte response.");
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/Frame.cs]: Too many bytes written

File: src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/Frame.cs [hunk 5]
```diff
@@ -548,6 +551,7 @@ public Task WriteAsync(ArraySegment<byte> data, CancellationToken cancellationTo
             }
 
             _responseBytesWritten += data.Count;
+            VerifyWrite();
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/Frame.cs]: this calling pattern is misleading, _responseBytesWritten gets incremented regardless of if you've written the bytes out. Invert it:
```
             VerifyWrite(data.Count);
             _responseBytesWritten += data.Count;
```

and/or move `_responseBytesWritten += data.Count;` inside of VerifyWrite and make it VerifyAndUpdateWrite

File: src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/FrameOfT.cs (hunk 3 above)
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/FrameOfT.cs]: Why even throw at this point? Why not just directly invoke ReportApplicationError?

File: src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/Frame.cs (hunk 4 above)
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/Frame.cs]: Yes, Stream allows InvalidOperationException. 
https://msdn.microsoft.com/en-us/library/hh137799(v=vs.110).aspx

File: src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/FrameOfT.cs (hunk 3 above)
Comment (cesarbs) [inline src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/FrameOfT.cs]: Why have two different messages when we can have a single one that reports the error in the same way?

File: src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/Frame.cs (hunk 1 above)
Comment (cesarbs) [inline src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/Frame.cs]: I'll leave it as `Parse` since that will throw if the app does something like `                httpContext.Response.Headers["Content-Length"] = "foo";`

**Comment by cesarbs** [comment]
@Tratcher I've addressed your comments.

File: src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/Frame.cs [hunk 6]
```diff
@@ -598,6 +602,32 @@ public Task WriteAsync(ArraySegment<byte> data, CancellationToken cancellationTo
...
+        }
+
+        protected void TryInitializeResponseContentLength()
+        {
+            if (!_responseContentLength.HasValue)
+            {
+                var responseHeaders = FrameResponseHeaders;
+                if (responseHeaders != null && responseHeaders.HasContentLength)
+                {
+                    _responseContentLength = long.Parse(responseHeaders.HeaderContentLength.ToString(), NumberStyles.None);
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/Frame.cs]: This will result in a FormatException being thrown from Write. Use TryParse and ignore values you can't parse.
When determining if you should chunk, do you check for the presence of a content-length or if you understand it?

Comment (cesarbs) [inline src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/Frame.cs]: We check for its presence but we don't try to parse it.

I disagree with you on this point. We should throw if the `Content-Length` header is set to an invalid value. The whole point of this change is to track errors related to response `Content-Length`, this would be one of them. Without a valid value, we can't track it.

Comment (Tratcher) [inline src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/Frame.cs]: Even if you want to throw, you can't throw a FormatException from Stream.Write. It needs to be wrapped in an InvalidOperationException

File: src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/Frame.cs (hunk 1 above)
Comment (halter73) [inline src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/Frame.cs]: In that case you should throw from the header setter ideally.

I think we should make sure we have the [same logic we use in HttpAbstractions](https://github.com/aspnet/HttpAbstractions/blob/45696535048559188dce8b38a9fde82baa88893c/src/Microsoft.AspNetCore.Http/Internal/ParsingHelpers.cs#L407).  We should probably look into sharing this with HttpAbsractions somehow.

It's also a good opportunity to make sure we use the same logic for parsing the request header content-length. Right now, it seems like we're probably more lax than we should be.

Comment (halter73) [inline src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/Frame.cs]: Oh. And I see you changed this to a long 👍 Were there not any tests failing when this was int.Parse?

**Comment by cesarbs** [comment]
Added validation to header setter. Not sure I covered all bases though. I'm thinking of adding a `long _contentLength` field to `FrameHeaders`, this would give fast access to that value instead of having to parse it from `HeaderContentLength`.

File: src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/FrameHeaders.Generated.cs [hunk 7]
```diff
@@ -201,6 +202,7 @@ public StringValues HeaderContentLength
             }
             set
             {
+                ValidateLongValue(value);
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/FrameHeaders.Generated.cs]: can you cache the parsed value internally rather than re-parse it later?

File: src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/Frame.cs (hunk 1 above)
Comment (halter73) [inline src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/Frame.cs]: Wow. I thought our LargeDownload test wrote over 2GB like the LargeUpload test. It looks like the LargeDownload test only has a 1MB body.

We should change the LargeDownload test to be like the LargeUpload test where we check every byte value in a few megabyte scenario, but still ensure that sending over 2GB works.

File: src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/FrameHeaders.cs [hunk 8]
```diff
@@ -232,6 +233,11 @@ public static void ValidateHeaderCharacters(string headerCharacters)
             }
         }
 
+        public static void ValidateLongValue(StringValues value)
+        {
+            long.Parse(value, NumberStyles.AllowLeadingWhite | NumberStyles.AllowTrailingWhite, CultureInfo.InvariantCulture);
```
Comment (halter73) [inline src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/FrameHeaders.cs]: Might make things a little easier to debug if this was wrapped in an InvalidOperationException stating the Content-Length header must be a positive decimal number or something like this.

File: src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/FrameHeaders.Generated.cs (hunk 7 above)
Comment (cesarbs) [inline src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/FrameHeaders.Generated.cs]: Yeah, it's what I was referring to in my comment above.

**Comment by cesarbs** [comment]
Updated. Caching Content-Length and wrapping `FormatException` in `InvalidOperationException`.

File: src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/FrameHeaders.Generated.cs [hunk 9]
```diff
@@ -12,6 +13,7 @@ public partial class FrameRequestHeaders
 
         private long _bits = 0;
         private HeaderReferences _headers;
+        private long? _contentLength;
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/FrameHeaders.Generated.cs]: Does this get cleared if I call `Headers.Remove("Content-Length")`?

Comment (cesarbs) [inline src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/FrameHeaders.Generated.cs]: Done.

**Comment by cesarbs** [comment]
I'm applying the `FrameHeaders` change to `FrameResponseHeaders` only (https://github.com/aspnet/KestrelHttpServer/pull/1155/commits/1d3b71d6ab8442ce064879d2bbe6dac37192f34a)

File: test/Microsoft.AspNetCore.Server.Kestrel.FunctionalTests/ResponseTests.cs [hunk 10]
```diff
@@ -324,7 +326,279 @@ public class ResponseTests
...
+                return TaskCache.CompletedTask;
+            }, serviceContext))
+            {
+                using (var connection = server.CreateConnection())
+                {
+                    await connection.Send(
+                        "GET / HTTP/1.1",
+                        "",
+                        "");
+                    await connection.ReceiveEnd(
```
Comment (Tratcher) [inline test/Microsoft.AspNetCore.Server.Kestrel.FunctionalTests/ResponseTests.cs]: Would this result in a RST? or just a FIN?

File: tools/Microsoft.AspNetCore.Server.Kestrel.GeneratedCode/KnownHeaders.cs [hunk 11]
```diff
@@ -349,7 +361,8 @@ protected override bool RemoveFast(string key)
                         if (""{header.Name}"".Equals(key, StringComparison.OrdinalIgnoreCase))
                         {{
                             if ({header.TestBit()})
-                            {{
+                            {{{If(header.Identifier == "ContentLength", () => $@"
```
Comment (halter73) [inline tools/Microsoft.AspNetCore.Server.Kestrel.GeneratedCode/KnownHeaders.cs]: Missed `loop.ClassName == "FrameResponseHeaders" && ` here.

File: tools/Microsoft.AspNetCore.Server.Kestrel.GeneratedCode/KnownHeaders.cs [hunk 12]
```diff
@@ -369,6 +382,7 @@ protected override void ClearFast()
         {{
             _bits = 0;
             _headers = default(HeaderReferences);
+            _contentLength = null;
```
Comment (halter73) [inline tools/Microsoft.AspNetCore.Server.Kestrel.GeneratedCode/KnownHeaders.cs]: Missed If here too.

File: test/Microsoft.AspNetCore.Server.Kestrel.FunctionalTests/ResponseTests.cs (hunk 10 above)
Comment (cesarbs) [inline test/Microsoft.AspNetCore.Server.Kestrel.FunctionalTests/ResponseTests.cs]: FIN

File: src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/Frame.cs [hunk 13]
```diff
@@ -684,7 +703,7 @@ private void ProduceStart(bool appCompleted)
 
         protected Task TryProduceInvalidRequestResponse()
         {
-            if (_requestRejected)
+            if (_requestRejected || _applicationException != null)
```
Comment (halter73) [inline src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/Frame.cs]: Why is this change necessary?

Comment (halter73) [inline src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/Frame.cs]: Obviously the application could swallow the exception, so we shouldn't be relying on it. And this change has little to do with invalid requests, right? I guess we're now parsing the request content-length more strictly...

Comment (cesarbs) [inline src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/Frame.cs]: If the response hasn't already started by the time we detect a mismatch in the content length, we can send a 500 response. `_applicationException` is set when there's a mismatch, so we need this check to run the error response logic.

Comment (halter73) [inline src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/Frame.cs]: That should be handled by the normal call to `ProduceEnd()`.

**Comment by halter73** [comment]
I don't think we need a new event id for this. Treating it as a normal application error is fine with me.

File: src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/Frame.cs (hunk 13 above)
Comment (cesarbs) [inline src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/Frame.cs]: Indeed. My very first change (that I didn't even commit) was calling `Abort()` on a content length mismatch. This would prevent the call to `ProduceEnd()` in `FrameOfT`. But since I moved away from that decision, `ProduceEnd()` handles it as expected.

**Comment by cesarbs** [comment]
Updated.

**Comment by halter73** [comment]
:shipit:

File: tools/Microsoft.AspNetCore.Server.Kestrel.GeneratedCode/KnownHeaders.cs [hunk 14]
```diff
@@ -228,12 +235,14 @@ public partial class {loop.ClassName}
...
             set
-            {{
+            {{{If(loop.ClassName == "FrameResponseHeaders" && header.Identifier == "ContentLength", () => $@"
+                _contentLength = ParseContentLength(value);")}
                 {header.SetBit()};
                 _headers._{header.Identifier} = value; {(header.EnhancedSetter == false ? "" : $@"
                 _headers._raw{header.Identifier} = null;")}
             }}
         }}")}
+        {(loop.ClassName == "FrameResponseHeaders" ? "public long? HeaderContentLengthValue => _contentLength;" : "")}
```
Comment (halter73) [inline tools/Microsoft.AspNetCore.Server.Kestrel.GeneratedCode/KnownHeaders.cs]: nit: I would probably put this in FrameResponseHeaders.cs

File: tools/Microsoft.AspNetCore.Server.Kestrel.GeneratedCode/KnownHeaders.cs [hunk 15]
```diff
@@ -199,6 +204,7 @@ public static string GeneratedFile()
             return $@"
 using System;
 using System.Collections.Generic;
+using System.Globalization;
```
Comment (halter73) [inline tools/Microsoft.AspNetCore.Server.Kestrel.GeneratedCode/KnownHeaders.cs]: nit: I don't think this is being used.

File: tools/Microsoft.AspNetCore.Server.Kestrel.GeneratedCode/KnownHeaders.cs [hunk 16]
```diff
@@ -304,7 +313,8 @@ protected override void SetValueFast(string key, StringValues value)
                 case {byLength.Key}:
                     {{{Each(byLength, header => $@"
                         if (""{header.Name}"".Equals(key, StringComparison.OrdinalIgnoreCase))
-                        {{
+                        {{{If(loop.ClassName == "FrameResponseHeaders" && header.Identifier == "ContentLength", () => $@"
```
Comment (halter73) [inline tools/Microsoft.AspNetCore.Server.Kestrel.GeneratedCode/KnownHeaders.cs]: nit: `$` is unnecessary.

File: tools/Microsoft.AspNetCore.Server.Kestrel.GeneratedCode/KnownHeaders.cs [hunk 17]
```diff
@@ -328,7 +338,9 @@ protected override void AddValueFast(string key, StringValues value)
                             if ({header.TestBit()})
                             {{
                                 ThrowDuplicateKeyException();
-                            }}
+                            }}{
+                            If(loop.ClassName == "FrameResponseHeaders" && header.Identifier == "ContentLength", () => $@"
+                            _contentLength = ParseContentLength(value);")}
```
Comment (halter73) [inline tools/Microsoft.AspNetCore.Server.Kestrel.GeneratedCode/KnownHeaders.cs]: No `$` needed

File: tools/Microsoft.AspNetCore.Server.Kestrel.GeneratedCode/KnownHeaders.cs [hunk 18]
```diff
@@ -228,12 +235,14 @@ public partial class {loop.ClassName}
                 return StringValues.Empty;
             }}
             set
-            {{
+            {{{If(loop.ClassName == "FrameResponseHeaders" && header.Identifier == "ContentLength", () => $@"
```
Comment (halter73) [inline tools/Microsoft.AspNetCore.Server.Kestrel.GeneratedCode/KnownHeaders.cs]: No `$`.

File: tools/Microsoft.AspNetCore.Server.Kestrel.GeneratedCode/KnownHeaders.cs [hunk 19]
```diff
@@ -349,7 +361,8 @@ protected override bool RemoveFast(string key)
                         if (""{header.Name}"".Equals(key, StringComparison.OrdinalIgnoreCase))
                         {{
                             if ({header.TestBit()})
-                            {{
+                            {{{If(loop.ClassName == "FrameResponseHeaders" && header.Identifier == "ContentLength", () => $@"
```
Comment (halter73) [inline tools/Microsoft.AspNetCore.Server.Kestrel.GeneratedCode/KnownHeaders.cs]: ditto

**Comment by cesarbs** [comment]
@halter73 nits addressed :)
